llama-index-core/llama_index/core/workflow/context.py

```diff
@@ -242,7 +247,53 @@ def send_event(self, message: Event, step: Optional[str] = None) -> None:
                     f"Step {step} does not accept event of type {type(message)}"
                 )
 
-        self._broker_log.append(message)
+    def _create_checkpoint(
+        self,
+        last_completed_step: Optional[str],
+        input_ev: Optional[Event],
+        output_ev: Event,
+    ) -> None:
+        """Build a checkpoint around the last completed step."""
+        checkpoint = Checkpoint(
+            last_completed_step=last_completed_step,
+            input_event=input_ev,
+            output_event=output_ev,
+            ctx_state=self.to_dict(),
```

Comment:
to_dict takes a serializer parameter (for which we let users specify json, json with pickle fallback, or their own subclassed serializer) -- do we want to expose that?

Reply:
Yeah, I think we should expose this. I guess we need to expose this somehow to user at the `workflow` level too so that they can configure the serializer. Maybe in the `run(context_serializer: Optional[...] = None)` method?

Reply:
I'd probably pass it to `__init__` to avoid having different types of checkpoint for different runs in the same workflow instance.

Reply:
Added `checkpoint_serializer` param to `__init__` which is used to assign a new private attr `_checkpoint_serializer`. I'm not 100% sold on this name, but can't think of a better alternative at this point.